pkgs/development/ruby-modules/bundler-env/default.nix

```diff
@@ -67,25 +76,54 @@ let
     if hasBundler then gems.bundler
     else defs.bundler.override (attrs: { inherit ruby; });
 
-  gems = lib.flip lib.mapAttrs configuredGemset (name: attrs:
-    buildRubyGem ((removeAttrs attrs ["source"]) // attrs.source // {
-      inherit ruby;
-      gemName = name;
-      gemPath = map (gemName: gems."${gemName}") (attrs.dependencies or []);
-    }));
+  pathDerivation = {
+    usesGemspec ? false, ...
+  }@attrs:
+    let
+      res = {
+          inherit usesGemspec;
+          type = "derivation";
+          name = attrs.gemName;
+          version = attrs.version;
+          outPath = attrs.path;
+          outputs = [ "out" ];
+          out = res;
+          outputName = "out";
+        };
+    in res;
+
+  buildGem = name: attrs: (
+    let
+      gemAttrs = ((removeAttrs attrs ["source"]) // attrs.source // {
+        inherit ruby;
+        gemName = name;
+        gemPath = map (gemName: gems."${gemName}") (attrs.dependencies or []);
+      });
+    in
+    if gemAttrs.type == "path" then pathDerivation gemAttrs
+    else buildRubyGem gemAttrs);
+
+  gems = lib.flip lib.mapAttrs configuredGemset (name: attrs: buildGem name attrs);
+
+  maybeCopyAll = { usesGemspec ? false, ...}@main:
+  (if usesGemspec then ''
+  cp -a ${gemdir}/* $out/
+  '' else ""
+  );
 
   # We have to normalize the Gemfile.lock, otherwise bundler tries to be
   # helpful by doing so at run time, causing executables to immediately bail
   # out. Yes, I'm serious.
   confFiles = runCommand "gemfile-and-lockfile" {} ''
     mkdir -p $out
-    cp ${gemfile'} $out/Gemfile
-    cp ${lockfile'} $out/Gemfile.lock
+    ${maybeCopyAll mainGem}
+    cp ${gemfile'} $out/Gemfile || ls -l $out/Gemfile
+    cp ${lockfile'} $out/Gemfile.lock || ls -l $out/Gemfile.lock
   '';
 
   envPaths = lib.attrValues gems ++ lib.optional (!hasBundler) bundler;
 
-  binPaths = if mainGem != null then [ mainGem ] else envPaths;
+  binPaths = if mainGem != null then [ mainGem ] ++ envPaths else envPaths;
```

Comment:
Yeah we shouldn't link any binaries unless they are part of the main gem.

There is a bit of transition issue because the `pname` attribute wasn't there initially so it was not possible to know what was the main gem.